subsys/cpp/cpp_guard.c

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2019 Erwin Rol <erwin@erwinrol.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <sys/__assert.h>
+
+static K_MUTEX_DEFINE(guard_mutex);
```

Comment:
User mode code won't be able to use this `k_mutex`. `sys_mutex` would allow this however.

Is there a specific need for priority inheritance? If not I think a semaphore would be better, it will pull in less code on small systems. A `sys_sem` can also be used from user mode if it is placed in the default memory domain.

I suggest:
```
#include <app_memory/app_memdomain.h>

static K_APP_DMEM(z_libc_partition) SYS_SEM_DEFINE(heap_sem, 1, 1);
```

And use `sys_sem_take()`/`sys_sem_give()` to implement the critical section.

Reply:
@andrewboie 

> Is there a specific need for priority inheritance?

Hmmm it is a "global" lock so it think priority inversion could happen. If thread A starts the initialization of a static var and gets preempted before it is ready it will block thread B that might want to initialize another static var. And if than a thread with a prio in between A and B starts an endless loop, the high prio thread B never gets done. So unless I am thinking in the wrong direction, a mutex would be better ?

Reply:
I agree priority inversion could be a problem, but whatever's done this should probably use the same priority behavior as devices.  See #26073.